crates/sql_provider_datafusion/src/federation.rs

```diff
@@ -0,0 +1,83 @@
+use async_trait::async_trait;
+use datafusion_federation::{FederatedTableProviderAdaptor, FederatedTableSource};
+use datafusion_federation_sql::{SQLExecutor, SQLFederationProvider, SQLTableSource};
+use db_connection_pool::dbconnection::get_schema;
+use futures::TryStreamExt;
+use snafu::prelude::*;
+use std::sync::Arc;
+
+use crate::{get_stream, to_execution_error, SqlTable, UnableToGetSchemaSnafu};
+use datafusion::{
+    arrow::datatypes::SchemaRef,
+    error::{DataFusionError, Result as DataFusionResult},
+    physical_plan::{stream::RecordBatchStreamAdapter, SendableRecordBatchStream},
+    sql::{
+        sqlparser::dialect::{Dialect, GenericDialect},
+        TableReference,
+    },
+};
+
+impl<T, P> SqlTable<T, P> {
+    fn create_federated_table_source(
+        self: Arc<Self>,
+    ) -> DataFusionResult<Arc<dyn FederatedTableSource>> {
+        let table_name = self.table_reference.table().to_string();
+        let schema = Arc::clone(&self.schema);
+        let fed_provider = Arc::new(SQLFederationProvider::new(self));
+        Ok(Arc::new(SQLTableSource::new_with_schema(
+            fed_provider,
+            table_name,
+            schema,
+        )?))
+    }
+
+    pub fn create_federated_table_provider(
+        self: Arc<Self>,
+    ) -> DataFusionResult<FederatedTableProviderAdaptor> {
+        let table_source = Self::create_federated_table_source(Arc::clone(&self))?;
+        Ok(FederatedTableProviderAdaptor::new_with_provider(
+            table_source,
+            self,
+        ))
+    }
+}
+
+#[async_trait]
+impl<T, P> SQLExecutor for SqlTable<T, P> {
```

Comment:
This is the trait (`SQLExecutor`) that needs to be implemented which can turn something into a federated table. The main method is `execute` which takes a query string and returns a `SendableRecordBatchStream`.